### DISCUSSION TITLE
Allow Calling Extension Methods on Methods

### DISCUSSION THREAD

**Discussion by Jlobblet** [dotnet/csharplang#6174] [Language Ideas]
You can make an extension method on a `Func`, but it can't be called on a method (unless you cast it first). The proposal is illustrated by this snippet:

```cs
using System;

public static class Program
{
	public static T Hi<T>(this Func<T> func)
	{
		Console.WriteLine("hi!");
		return func();
	}
	
	public static int GetThree() => 3;
	
	public static void Main()
	{
		Console.WriteLine(GetThree.Hi()); // Note here

	}
}
```

Allowing this would make higher order functions that wrap the behaviour of another function far more fluid to write in chains. At the moment, `Hi(GetThree)` (or `((Func<int>)GetThree).Hi()`) works fine, but I think for longer method chains this is a bit nasty.

For a motivating example, consider a Result type such as F# or Rust's. The verison presented here is basic to avoid it being obstructive to the demonstration.
```cs
using System;

public class Result<T, TErr>
{
    public bool IsOk { get; init; }
    public T Value { get; init; }
    public TErr ErrValue { get; init; }
}

public static class ResultExtensions
{
    public static Result<T, Exception> Protect<T>(this Func<T> func)
    {
        try
        {
            return new Result<T, Exception>() { IsOk = true, Value = func() };
        }
        catch (Exception e)
        {
            return new Result<T, Exception>() { IsOk = false, ErrValue = e };
        }
    }
	
	
    public static Result<TReturn, Exception> Protect<T1, TReturn>(this Func<T1, TReturn> func, T1 t1) { /* etc. */ }
    public static Result<TReturn, Exception> Protect<T1, T2, TReturn>(this Func<T1, T2, TReturn> func, T1 t1, T2 t2) { /* etc. */ }
    public static Result<TReturn, Exception> Protect<T1, T2, T3, TReturn>(this Func<T1, T2, T3, TReturn> func, T1 t1, T2 t2, T3 t3) { /* etc. */ }
}
```

Which would allow for code such as:
```cs
File.ReadAllText.Protect("foo.txt");
File.Open.Protect("bar.txt", FileMode.Append);
```
and so on.

I think it would be a rich addition to the language, unlocking a lot of possiblities.

**Comment by Rekkonnect** [comment]
The only issue with this is the conflicting types. Other than that, I like the idea of being able to modify/interact with your delegate instance that you can quickly create with a method group through an extension.

There's two main issues about ambiguous types:
- Method group contains overloads whose signatures match those of delegate types that also have extensions with the same name, for example:
```csharp
private void Anything(int a);
private void Anything(int a, int b);

public static void DoAnything(this Action<int> action);
public static void DoAnything(this Action<int, int> action);

Anything.DoAnything(); // Action<int> or Action<int, int>
```
- A similar to the above conflict case, multiple delegate types having extensions with the same signature, where the delegate definitions have the same signature, for example:
```csharp
private void Anything(int a);

public static void DoAnything(this Action<int> action);
public static void DoAnything(this AnythingAction action);

public delegate void AnythingAction(int a);

Anything.DoAnything(); // Action<int> or AnythingAction
```

**Reply by sab39** [reply]
Ambiguous definitions are already a thing, though:
```cs
public static void Ambiguous(this IComparable item) { }
public static void Ambiguous(this IConvertible item) { }

"a string".Ambiguous();
```
That ambiguity isn't considered a language limitation - when it happens it can be resolved either by a cast or sometimes by invoking the extension method as if it were a normal `static`. Is there any reason why the same couldn't apply for delegate types that can match multiple ways?
